### PR TITLE
Enabled masked for a bool tensor

### PR DESCRIPTION
**Pull request by izdeby**
Enabled masked methods them for a bool tensor.

### CODE REVIEW

**Comment by ezyang** [comment]
So, how exactly is the deprecation cycle going to work? Are we going to change the behavior of these functions to do uint8 indexing eventually?

**Comment by izdeby** [comment]
@ezyang, did you mean bool indexing? it works with uint8 now. 

So the current plan looks like this: 
1) add deprecation warnings for existing methods
2) [after release] replace byte masks (uint8) with bool. 

@gchanan, correct me if im wrong, please.

**Comment by gchanan** [comment]
it's not standard practice to deprecate things before the replacement is ready to use.  I don't see a replacement ready to use here?

**Comment by izdeby** [comment]
There are 2 options here:
1) enable bool masks right now which will work alongside with byte masks, add deprecation warning that byte masks will be gone
2) add deprecation warning that byte will be replaced with bool and do that after the release. 

This PR implements the 2nd option as discussed offline. Should i go with the 1st option instead?

**Comment by gchanan** [comment]
I must have misunderstood the options, then.  It's not useful to provide deprecation warnings without replacements.  As a user, there's nothing I can do with that information.

File: aten/src/ATen/Declarations.cwrap [hunk 1]
```diff
@@ -69,6 +69,8 @@
 [[
   name: _th_masked_fill_
   cname: maskedFill
+  cpu_bool: True
```
Comment (gchanan) [inline aten/src/ATen/Declarations.cwrap]: This allows you to have the value/return be a bool tensor, which I guess is fine, but doesn't really have anything to do with the other part of this PR, which is putting a deprecation warning on byte masks.  To enable a deprecation path you'd need to support the mask being a bool.

**Comment by izdeby** [comment]
Sure, no problems, will switch to option #1.

File: aten/src/TH/generic/THTensorEvenMoreMath.cpp [hunk 2]
```diff
@@ -67,10 +67,11 @@ void THTensor_(nonzero)(THLongTensor *subscript, THTensor *tensor)
   delete [] idx;
 }
 
-#if !defined(TH_REAL_IS_BOOL) /* non bool only part */
-
 void THTensor_(maskedFill)(THTensor *tensor, THByteTensor *mask, scalar_t value)
 {
+  AT_WARN("Please, note that this behavior is about to change. " \
```
Comment (gchanan) [inline aten/src/TH/generic/THTensorEvenMoreMath.cpp]: to the above point about replacements, you want the deprecation warning to say what the user should do instead.  So, something like enable bool masks and then say something like:
"masked_fill_ received a mask with dtype torch.uint8, this behavior is now deprecated, please use dtype torch.bool for masks instead."

**Comment by izdeby** [comment]
@ezyang @gchanan, added option to pass bool masks to masked methods. Initially i wanted to add just overloaded versions but that turned out to be very hard to do and would require adding a new dispatcher which seems too much to me as byte masks will be killed after the release.

File: aten/src/ATen/Utils.h [hunk 3]
```diff
@@ -79,6 +79,20 @@ static inline TensorImpl* checked_tensor_unwrap(const Tensor& expr, const char *
   return expr.unsafeGetTensorImpl();
 }
 
+static inline TensorImpl* checked_bool_tensor_unwrap(const Tensor& expr, const char * name, int pos, bool allowNull, Backend backend, ScalarType scalar_type) {
```
Comment (gchanan) [inline aten/src/ATen/Utils.h]: why do you need this and why does it take a scalar_type?

File: aten/src/ATen/Declarations.cwrap [hunk 4]
```diff
@@ -69,35 +69,41 @@
...
```
Comment (gchanan) [inline aten/src/ATen/Declarations.cwrap]: can you separate this out in another PR (or just throw it away?).  This has nothing to do with masking, you are just enabling these functions for bool values, but they don't support half values (what's the difference?) and there is no larger audit or what should be supported (presumably index_ functions should also be supported?  i.e. everything that doesn't support masking).

File: aten/src/TH/generic/THTensorEvenMoreMath.cpp [hunk 5]
```diff
@@ -67,10 +67,13 @@ void THTensor_(nonzero)(THLongTensor *subscript, THTensor *tensor)
   delete [] idx;
 }
 
-#if !defined(TH_REAL_IS_BOOL) /* non bool only part */
-
-void THTensor_(maskedFill)(THTensor *tensor, THByteTensor *mask, scalar_t value)
+void THTensor_(maskedFill)(THTensor *tensor, THBoolTensor *mask, scalar_t value)
 {
+  // TODO: remove ScalarType::Byte path after the release as deprecated.
+  if (mask->dtype() == at::ScalarType::Byte) {
```
Comment (gchanan) [inline aten/src/TH/generic/THTensorEvenMoreMath.cpp]: this is pretty confusing -- why does this take a THBoolTensor but that can be a byte?  Instead, why not just write overloads (or give them different names, whatever works better), for Bool and Byte versions?  They can call into the same implementation after the checking is done.

File: aten/src/THC/generic/THCTensorMasked.cu [hunk 6]
```diff
@@ -2,38 +2,48 @@
...
-  THCAssertSameGPU(THCTensor_(checkGPU)(state, 2, tensor, mask));
-  THArgCheck(THCTensor_(nElement)(state, tensor) ==
-             THCudaByteTensor_nElement(state, mask),
-             2, "sizes do not match");
-
-  if (!THC_pointwiseApply2<scalar_t, uint8_t>(state, tensor, mask,
-                                          TensorMaskedFillOp<scalar_t, unsigned char>(value))) {
-    THArgCheck(false, 2, CUTORCH_DIM_WARNING);
+  //TODO: clean up the byte mask path after the release.
+  if (mask->dtype() == at::ScalarType::Byte) {
```
Comment (gchanan) [inline aten/src/THC/generic/THCTensorMasked.cu]: same here.

File: aten/src/THC/generic/THCTensorMasked.cu [hunk 7]
```diff
@@ -2,38 +2,48 @@
...
-  THCTensor_(maskedFill)(state, tensor, maskCuda, value);
-  THCudaByteTensor_free(state, maskCuda);
-}
-
 void THCTensor_(maskedCopy)(THCState* state,
-                            THCTensor *tensor, THCudaByteTensor *mask, THCTensor *src)
+                            THCTensor *tensor, THCudaBoolTensor *mask, THCTensor *src)
 {
+  //TODO: clean up the byte mask path after the release.
+  if (mask->dtype() == at::ScalarType::Byte) {
```
Comment (gchanan) [inline aten/src/THC/generic/THCTensorMasked.cu]: same here.

File: aten/src/THC/generic/THCTensorMasked.cu [hunk 8]
```diff
@@ -95,24 +114,28 @@ void THCTensor_(maskedCopy)(THCState* state,
...
-
 void THCTensor_(maskedSelect)(THCState* state,
-                              THCTensor* tensor, THCTensor* src, THCudaByteTensor* mask) {
-  THCAssertSameGPU(THCTensor_(checkGPU)(state, 3, tensor, src, mask));
-  THArgCheck(THCudaByteTensor_nElement(state, mask) ==
-             THCTensor_(nElement)(state, src),
-             2, "sizes do not match");
+                              THCTensor* tensor, THCTensor* src, THCudaBoolTensor* mask) {
+  //TODO: clean up the byte mask path after the release.
+  if (mask->dtype() == at::ScalarType::Byte) {
```
Comment (gchanan) [inline aten/src/THC/generic/THCTensorMasked.cu]: same here.

File: test/test_torch.py [hunk 9]
```diff
@@ -7595,6 +7618,17 @@ def test_masked_fill(self):
         dst2.masked_fill_(dst2 > 0, val)
         self.assertEqual(dst, dst2, 0)
 
+        # test bool tensor case
+        for device in torch.testing.get_all_device_types():
+            x = torch.tensor([True, True, False, False], dtype=torch.bool, device=device)
```
Comment (gchanan) [inline test/test_torch.py]: can you check that there is a warning with byte tensors?  Presumably you also want to guard any existing calls in tests that may exist so we don't spew warnings.

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/TH/generic/THTensorEvenMoreMath.cpp (hunk 5 above)
Comment (izdeby) [reply aten/src/TH/generic/THTensorEvenMoreMath.cpp]: Initially i wanted to add an overload version which would be taking THBoolTensor mask, but no matter which mask i would pass, bool or byte, it will still go through the THByteTensor version and fail due to storage type errors. Long story short, @ezyang, helped me figure out that there is a dispatcher which decides which overload to use and in order to make it work with overloads, we have to add a new dispatcher. This solution seems a bit too complex for a given problem assuming that Byte masks will be killed soon. 

So thats why i've added checked_bool_tensor_unwrap which has to work for Bool and Byte types.

**Comment by ezyang** [comment]
@gchanan I'll let you review this, but let me know if you want me to look (I helped Iurii debug some issues while he was working on this patch.)

File: aten/src/ATen/Declarations.cwrap [hunk 10]
```diff
@@ -84,6 +84,24 @@
...
 ]]
+[[
+  name: _th_masked_fill_bool_
+  cname: maskedFillBool
+  variants: function
+  return: self
+  options:
+    - arguments:
+      - arg: THTensor* self
+        broadcast: mask inplace fallback types:Byte
```
Comment (gchanan) [inline aten/src/ATen/Declarations.cwrap]: this "types:Byte" thing looks questionable?

File: aten/src/ATen/Declarations.cwrap [hunk 11]
```diff
@@ -84,6 +84,24 @@
...
+      - THBoolTensor* mask
+      - real value
+    - zero_dim_tensor_only: True
+      arguments:
+      - arg: THTensor* self
+        broadcast: mask inplace fallback types:Byte
```
Comment (gchanan) [inline aten/src/ATen/Declarations.cwrap]: here too.

File: aten/src/TH/generic/THTensorEvenMoreMath.cpp [hunk 12]
```diff
@@ -67,6 +67,26 @@ void THTensor_(nonzero)(THLongTensor *subscript, THTensor *tensor)
   delete [] idx;
 }
 
+accreal THTensor_(sumall)(THTensor *tensor)
```
Comment (gchanan) [inline aten/src/TH/generic/THTensorEvenMoreMath.cpp]: why do you need these?

File: aten/src/ATen/function_wrapper.py [hunk 13]
```diff
@@ -215,6 +216,7 @@ def __init__(self, reason):
     'THTensor*': 'Tensor',
     'THSTensor*': 'SparseTensorRef',
     'THByteTensor*': 'BoolTensor',
+    'THBoolTensor*': 'BoolTensor',
```
Comment (gchanan) [inline aten/src/ATen/function_wrapper.py]: wait both Bool and Byte map to BoolTensor?  How do you check you have the properly dtype at the TH boundary then?

**Review by gchanan — COMMENTED** [review]
why do you need sum ?

File: aten/src/TH/generic/THTensorEvenMoreMath.cpp (hunk 12 above)
Comment (izdeby) [reply aten/src/TH/generic/THTensorEvenMoreMath.cpp]: We are using them in THTensor_(maskedSelectBool) for resizing the result tensor.
Example with byte tensor: https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THTensorEvenMoreMath.cpp#L137

File: aten/src/ATen/native/LegacyDefinitions.cpp [hunk 14]
```diff
@@ -51,15 +51,33 @@ Tensor& zero_(Tensor& self) {
 }
 
 Tensor & masked_fill_(Tensor& self, const Tensor & mask, Scalar value) {
-  return at::legacy::th::_th_masked_fill_(self, mask, value);
+  if (mask.dtype() == at::ScalarType::Byte) {
+    AT_WARN("masked_fill_ received a mask with dtype torch.uint8, this behavior is now deprecated," \
```
Comment (gchanan) [inline aten/src/ATen/native/LegacyDefinitions.cpp]: I would avoid putting in the warnings for now until we change the default of comparison ops -- otherwise you force users to write unnecessary `.to(torch.bool)` in a number of places.

File: test/test_torch.py [hunk 15]
```diff
@@ -3809,11 +3809,11 @@ def tensorfn(myfn, t1, t2):
                     if fn == "lerp":
                         return myfn(t1, 0.5)
                     elif fn == "masked_select":
-                        return myfn(t1 < 0)
+                        return myfn((t1 < 0).to(torch.bool))
```
Comment (gchanan) [inline test/test_torch.py]: As above, I'd avoid these changes because the comparison op change will handle this for free.

File: aten/src/ATen/function_wrapper.py [hunk 16]
```diff
@@ -773,7 +786,7 @@ def find_dispatch_tensor(formals):
         # type: (List[AtFormal]) -> Optional[str]
         # dispatch to self if it's a parameter
         def is_any_tensor_type(formal):
-            return (formal['dynamic_type'] == 'Tensor' or formal['dynamic_type'] == 'BoolTensor'
+            return (formal['dynamic_type'] == 'Tensor' or formal['dynamic_type'] == 'ByteTensor'
```
Comment (gchanan) [inline aten/src/ATen/function_wrapper.py]: this looks like it should have an `or 'BoolTensor'`.

File: aten/src/ATen/native/Itertools.cpp [hunk 17]
```diff
@@ -51,7 +51,7 @@ Tensor combinations(const Tensor& self, int64_t r, bool with_replacement) {
   std::vector<Tensor> grids = at::meshgrid(std::vector<Tensor>(r, self));
   Tensor mask = _triu_mask(num_elements, r, with_replacement, self.options());
   for(Tensor &t : grids) {
-    t = t.masked_select(mask);
+    t = t.masked_select(mask.to(at::kBool));
```
Comment (gchanan) [inline aten/src/ATen/native/Itertools.cpp]: nit: you don't need this.

File: aten/src/ATen/native/LegacyDefinitions.cpp [hunk 18]
```diff
@@ -51,15 +51,27 @@ Tensor& zero_(Tensor& self) {
 }
 
 Tensor & masked_fill_(Tensor& self, const Tensor & mask, Scalar value) {
-  return at::legacy::th::_th_masked_fill_(self, mask, value);
+  if (mask.dtype() == at::ScalarType::Byte) {
```
Comment (gchanan) [inline aten/src/ATen/native/LegacyDefinitions.cpp]: nit: a comment about why you need to do this would be useful (just say something like we dispatch on `self` and TH is type-chcked, so we need different definitions.  This can be fixed by moving to ATen`.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 19]
```diff
@@ -401,7 +401,7 @@ Tensor& logsumexp_out(Tensor& result, const Tensor &self, IntArrayRef dims, bool
   if (self.numel() != 0) {
     auto maxes = at::max_values(self, dims, true);
     auto maxes_squeezed = (keepdim ? maxes : squeeze_multiple(maxes, dims));
-    maxes_squeezed.masked_fill_(maxes_squeezed.abs() == INFINITY, 0);
+    maxes_squeezed.masked_fill_((maxes_squeezed.abs() == INFINITY).to(at::kBool), 0);
```
Comment (gchanan) [inline aten/src/ATen/native/ReduceOps.cpp]: nit: you don't need this.

File: tools/autograd/gen_variable_type.py [hunk 20]
```diff
@@ -483,7 +483,7 @@ def is_differentiable(arg):
             return False
         if 'Tensor' not in arg['type']:
             return False
-        if arg['dynamic_type'] in {'IndexTensor', 'BoolTensor'}:
+        if arg['dynamic_type'] in {'IndexTensor', 'ByteTensor'}:
```
Comment (gchanan) [inline tools/autograd/gen_variable_type.py]: this should also include BoolTensor?

File: tools/autograd/gen_variable_type.py [hunk 21]
```diff
@@ -631,7 +631,7 @@ def emit_check_no_requires_grad(tensor_args, args_with_derivatives):
                 # Double-backwards definitions sometimes take in 'input' and
                 # 'output', but only define the derivative for input.
                 continue
-            if arg['dynamic_type'] in {'IndexTensor', 'BoolTensor'}:
+            if arg['dynamic_type'] in {'IndexTensor', 'ByteTensor'}:
```
Comment (gchanan) [inline tools/autograd/gen_variable_type.py]: here too?

File: torch/_tensor_str.py [hunk 22]
```diff
@@ -84,7 +84,7 @@ def __init__(self, tensor):
                 self.max_width = max(self.max_width, len(value_str))
 
         else:
-            nonzero_finite_vals = torch.masked_select(tensor_view, torch.isfinite(tensor_view) & tensor_view.ne(0))
+            nonzero_finite_vals = torch.masked_select(tensor_view, (torch.isfinite(tensor_view) & tensor_view.ne(0)).to(torch.bool))
```
Comment (gchanan) [inline torch/_tensor_str.py]: nit: you don't need this.

File: test/test_torch.py [hunk 23]
```diff
@@ -7600,6 +7600,14 @@ def test_masked_scatter(self):
         with self.assertRaises(RuntimeError):
             dest.masked_scatter_(mask, src)
 
+        # test bool tensor case
+        for device in torch.testing.get_all_device_types():
```
Comment (gchanan) [inline test/test_torch.py]: I don't really understand the test -- I would assume you'd just run the same tests through bool and byte and check you get the same answers?

File: aten/src/ATen/native/LegacyDefinitions.cpp [hunk 24]
```diff
@@ -51,15 +51,33 @@ Tensor& zero_(Tensor& self) {
 }
 
 Tensor & masked_fill_(Tensor& self, const Tensor & mask, Scalar value) {
-  return at::legacy::th::_th_masked_fill_(self, mask, value);
+  // As we dispatch on self and TH is type-chcked, we need different definitions.
```
Comment (gchanan) [inline aten/src/ATen/native/LegacyDefinitions.cpp]: nit: type-checked.

File: aten/src/ATen/native/LegacyDefinitions.cpp [hunk 25]
```diff
@@ -51,15 +51,33 @@ Tensor& zero_(Tensor& self) {
...
+  // This can be fixed by moving to ATen.
+  if (mask.dtype() == at::ScalarType::Byte) {
+    return at::legacy::th::_th_masked_fill_(self, mask, value);
```
Comment (gchanan) [inline aten/src/ATen/native/LegacyDefinitions.cpp]: in a future PR, it might make sense to call the bool one masked_fill_ and the other one masked_fill_byte, since that is the end state, but not necessary for now.

File: aten/src/ATen/native/LegacyDefinitions.cpp [hunk 26]
```diff
@@ -51,15 +51,33 @@ Tensor& zero_(Tensor& self) {
...
+  } else {
+    return at::legacy::th::_th_masked_fill_bool_(self, mask, value);
+  }
 }
 
 Tensor & masked_fill_(Tensor& self, const Tensor & mask, const Tensor & value) {
-  return at::legacy::th::_th_masked_fill_(self, mask, value);
+  // As we dispatch on self and TH is type-chcked, we need different definitions.
```
Comment (gchanan) [inline aten/src/ATen/native/LegacyDefinitions.cpp]: type-checked.

File: aten/src/THC/generic/THCTensorMasked.cu [hunk 27]
```diff
@@ -165,6 +247,67 @@ void THCTensor_(maskedSelect)(THCState* state,
   THCudaCheck(cudaGetLastError());
 }
 
+void THCTensor_(maskedSelectBool)(THCState* state,
```
Comment (gchanan) [inline aten/src/THC/generic/THCTensorMasked.cu]: also for the future: I think you could move these definitions into a non-generic function so you don't have to copy the code.

**Review by gchanan — APPROVED** [review]
